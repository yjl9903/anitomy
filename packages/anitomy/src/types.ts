import type { ElementCategory } from './element';

export interface AnitomyOptions {
  delimiters: string;

  parseEpisodeNumber: boolean;

  parseEpisodeTitle: boolean;

  parseFileExtension: boolean;

  parseReleaseGroup: boolean;
}

export type ParsedResult = Partial<Record<ElementCategory, string>>;

export interface AnitomyResult {
  title: string | undefined;

  type: string | undefined;

  season: string | undefined;

  year: number | undefined;

  month: number | undefined;

  language: string | undefined;

  subtitles: string | undefined;

  source: string | undefined;

  episode: {
    number: number | undefined;

    numberAlt: number | undefined;

    title: string | undefined;
  };

  volume: {
    number: number | undefined;
  };

  video: {
    term: string | undefined;

    resolution: string | undefined;
  };

  audio: {
    term: string | undefined;
  };

  release: {
    version: number | undefined;
    group: string | undefined;
  };

  file: {
    name: string;

    checksum: string | undefined;

    extension: string | undefined;
  };

  // prefix: {
  //   season: string | undefined;
  //   volume: string | undefined;
  //   episode: string | undefined;
  // };
}
