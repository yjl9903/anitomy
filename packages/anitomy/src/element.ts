export enum ElementCategory {
  AnimeSeason = 'season',
  AnimeSeasonPrefix = 'prefix.season',
  AnimeTitle = 'title',
  AnimeType = 'type',
  AnimeYear = 'year',
  AnimeMonth = 'month', // Added by https://github.com/yjl9903/AnimeGarden
  DeviceCompatibility = 'DeviceCompatibility',
  Source = 'source',
  // Episode
  EpisodeNumber = 'episode.number',
  EpisodeNumberAlt = 'episode.numberAlt',
  EpisodePrefix = 'prefix.episode',
  EpisodeTitle = 'episode.title',
  // File
  FileChecksum = 'checksum',
  FileExtension = 'extension',
  FileName = 'filename',
  // Language
  Language = 'language',
  Subtitles = 'subtitles',
  // Audio and Video
  AudioTerm = 'audio.term',
  VideoResolution = 'video.resolution',
  VideoTerm = 'video.term',
  // Volume
  VolumeNumber = 'volume',
  VolumePrefix = 'prefix.volume',
  // Release
  ReleaseGroup = 'release.group',
  ReleaseInformation = 'release.information',
  ReleaseVersion = 'release.version',
  // Other
  Unknown = 'unknown',
  Other = 'other'
}
