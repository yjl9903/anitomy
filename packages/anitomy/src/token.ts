export class TextRange {
  public text: string;

  public offset: number;

  public size: number;

  public constructor(text: string, offset: number, size: number) {
    this.text = text;
    this.offset = offset;
    this.size = size;
  }

  public fork(offset: number, size: number) {
    return new TextRange(this.text, offset, size);
  }

  public toString() {
    return this.text.slice(this.offset, this.offset + this.size);
  }
}

export enum TokenCategory {
  Unknown = 'Unknown',
  Bracket = 'Bracket',
  Delimiter = 'Delimiter',
  Identifier = 'Identifier',
  Invalid = 'Invalid'
}

export enum TokenFlag {
  // None
  None,
  // Categories
  Bracket,
  NotBracket,
  Delimiter,
  NotDelimiter,
  Identifier,
  NotIdentifier,
  Unknown,
  NotUnknown,
  Valid,
  NotValid,
  // Enclosed (Meaning that it is enclosed in some bracket (e.g. [ ] ))
  Enclosed,
  NotEnclosed
}

export interface Token {
  category: TokenCategory;

  content: string;

  enclosed: boolean;
}

function checkTokenFlags(token: Token, flags: TokenFlag[]) {
  // Make sure token is the correct closure
  if (flags.some((f) => f === TokenFlag.Enclosed || f === TokenFlag.NotEnclosed)) {
    const success = flags.includes(TokenFlag.Enclosed) === token.enclosed;
    if (!success) return false; // Not enclosed correctly (e.g. enclosed when we're looking for non-enclosed).
  }

  // Make sure token is the correct category
  if (!flags.some((f) => TokenFlag.Bracket <= f && f <= TokenFlag.NotValid)) {
    return true;
  }

  const tasks: [TokenFlag, TokenFlag, TokenCategory][] = [
    [TokenFlag.Bracket, TokenFlag.NotBracket, TokenCategory.Bracket],
    [TokenFlag.Delimiter, TokenFlag.NotDelimiter, TokenCategory.Delimiter],
    [TokenFlag.Identifier, TokenFlag.NotIdentifier, TokenCategory.Identifier],
    [TokenFlag.Unknown, TokenFlag.NotUnknown, TokenCategory.Unknown],
    [TokenFlag.NotValid, TokenFlag.Valid, TokenCategory.Invalid]
  ];
  for (const [fe, fn, c] of tasks) {
    const success = flags.includes(fe)
      ? token.category === c
      : flags.includes(fn) && token.category !== c;
    if (success) return true;
  }
  return false;
}

export function findNextToken(tokens: Token[], position: number, ...flags: TokenFlag[]) {
  for (let i = position + 1; i < tokens.length; i++) {
    if (tokens[i] && checkTokenFlags(tokens[i], flags)) {
      return i;
    }
  }
  return tokens.length;
}

export function findPrevToken(tokens: Token[], position: number, ...flags: TokenFlag[]) {
  for (let i = position - 1; i >= 0; i--) {
    if (tokens[i] && checkTokenFlags(tokens[i], flags)) {
      return i;
    }
  }
  return -1;
}

export function findToken(tokens: Token[], start: number, end: number, ...flags: TokenFlag[]) {
  for (let i = start; i < end; i++) {
    if (tokens[i] && checkTokenFlags(tokens[i], flags)) {
      return i;
    }
  }
  return -1;
}
