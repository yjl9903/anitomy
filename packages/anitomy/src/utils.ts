import type { ParsedResult } from './types';

export function inRange<T>(list: T[], idx: number) {
  return 0 <= idx && idx < list.length;
}

export function isNumericString(text: string) {
  return /^\d+$/.test(text);
}

export function trim(text: string, removal: string[]) {
  let start = 0,
    end = text.length - 1;
  while (start <= end && removal.includes(text[start])) {
    start++;
  }
  while (end >= start && removal.includes(text[end])) {
    end--;
  }
  return text.slice(start, end + 1);
}

export function mergeResult(source: ParsedResult, income: ParsedResult = {}) {
  return {
    ...source,
    ...income
  };
}
