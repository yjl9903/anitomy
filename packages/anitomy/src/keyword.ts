import type { ParsedResult } from './types';

import { TextRange } from './token';
import { ElementCategory } from './element';

export interface KeywordOptions {
  identifiable: boolean;

  searchable: boolean;

  valid: boolean;
}

export type Keyword = {
  category: ElementCategory;
} & KeywordOptions;

const keys = new Map<string, Keyword>();

const extensions = new Map<string, Keyword>();

export class KeywordManager {
  private static keys = keys;

  private static extensions = extensions;

  private static peekEntries: Array<{ category: ElementCategory; list: string[] }> = [
    { category: ElementCategory.AudioTerm, list: ['Dual Audio'] },
    { category: ElementCategory.VideoTerm, list: ['H264', 'H.264', 'h264', 'h.264'] },
    {
      category: ElementCategory.VideoResolution,
      list: ['480p', '480P', '720p', '720P', '1080p', '1080P']
    },
    { category: ElementCategory.Source, list: ['Blu-Ray'] }
  ];

  static {
    const optionsDefault: KeywordOptions = { identifiable: true, searchable: true, valid: true };
    const optionsInvalid: KeywordOptions = { identifiable: false, searchable: false, valid: false };
    const optionsUnidentifiable: KeywordOptions = {
      identifiable: false,
      searchable: true,
      valid: true
    };
    const optionsUnidentifiableInvalid: KeywordOptions = {
      identifiable: false,
      searchable: true,
      valid: false
    };
    const optionsUnidentifiableUnsearchable: KeywordOptions = {
      identifiable: false,
      searchable: false,
      valid: true
    };

    // Season
    add(ElementCategory.AnimeSeasonPrefix, optionsUnidentifiable, ['SAISON', 'SEASON']);

    // Type
    add(ElementCategory.AnimeType, optionsUnidentifiable, [
      'GEKIJOUBAN',
      'MOVIE',
      'OAD',
      'OAV',
      'ONA',
      'OVA',
      'SPECIAL',
      'SPECIALS',
      'TV',
      '特别篇',
      '特別篇',
      '特別編',
      '特别话',
      '特別话',
      '特別話',
      '番外篇',
      '番外編'
    ]);
    add(ElementCategory.AnimeType, optionsUnidentifiableUnsearchable, ['SP']);
    add(ElementCategory.AnimeType, optionsUnidentifiableInvalid, [
      'ED',
      'ENDING',
      'NCED',
      'NCOP',
      'OP',
      'OPENING',
      'PREVIEW',
      'PV'
    ]);

    add(ElementCategory.AudioTerm, optionsDefault, [
      // Audio channels
      '2.0CH',
      '2CH',
      '5.1',
      '5.1CH',
      'DTS',
      'DTS-ES',
      'DTS5.1',
      'TRUEHD5.1',
      // Audio codec
      'AAC',
      'AACX2',
      'AACX3',
      'AACX4',
      'AC3',
      'EAC3',
      'E-AC-3',
      'FLAC',
      'FLACX2',
      'FLACX3',
      'FLACX4',
      'LOSSLESS',
      'MP3',
      'OGG',
      'VORBIS',
      // Audio language
      'DUALAUDIO',
      'DUAL AUDIO'
    ]);
    add(ElementCategory.VideoTerm, optionsDefault, [
      // Frame rate
      '23.976FPS',
      '24FPS',
      '29.97FPS',
      '30FPS',
      '60FPS',
      '120FPS',
      // Video codec
      '8BIT',
      '8-BIT',
      '10BIT',
      '10BITS',
      '10-BIT',
      '10-BITS',
      'HI10',
      'HI10P',
      'HI444',
      'HI444P',
      'HI444PP',
      'H264',
      'H265',
      'H.264',
      'H.265',
      'X264',
      'X265',
      'X.264',
      'AVC',
      'HEVC',
      'HEVC2',
      'HEVC-10BIT',
      'DIVX',
      'DIVX5',
      'DIVX6',
      'XVID',
      // Video format
      'AVI',
      'RMVB',
      'WMV',
      'WMV3',
      'WMV9',
      // Video quality
      'HQ',
      'LQ',
      // Video resolution
      'HD',
      'SD'
    ]);

    add(ElementCategory.Source, optionsDefault, [
      'BD',
      'BDRIP',
      'BLURAY',
      'BLU-RAY',
      'DVD',
      'DVD5',
      'DVD9',
      'DVD-R2J',
      'DVDRIP',
      'DVD-RIP',
      'R2DVD',
      'R2J',
      'R2JDVD',
      'R2JDVDRIP',
      'HDTV',
      'HDTVRIP',
      'TVRIP',
      'TV-RIP',
      'WEBCAST',
      'WEBDL',
      'WEB-DL',
      'WEBRIP',
      'WEB-RIP'
    ]);

    // Language
    add(ElementCategory.Language, optionsDefault, [
      'ENG',
      'ENGLISH',
      'ESPANO',
      'JAP',
      'PT-BR',
      'SPANISH',
      'VOSTFR',
      // feat: chinese
      'CHT',
      'CHS',
      '简中',
      '繁中',
      '简体',
      '繁體'
    ]);
    add(ElementCategory.Language, optionsUnidentifiable, ['ESP', 'ITA']);

    add(ElementCategory.Subtitles, optionsDefault, [
      'ASS',
      'GB', // feat: 简体字幕
      'BIG5',
      'DUB',
      'DUBBED',
      'HARDSUB',
      'HARDSUBS',
      'RAW',
      'SOFTSUB',
      'SOFTSUBS',
      'SUB',
      'SUBBED',
      'SUBTITLED'
    ]);

    // Episode
    add(ElementCategory.EpisodePrefix, optionsDefault, [
      'EP',
      'EP.',
      'EPS',
      'EPS.',
      'EPISODE',
      'EPISODE.',
      'EPISODES',
      'CAPITULO',
      'EPISODIO',
      'FOLGE'
    ]);
    add(ElementCategory.EpisodePrefix, optionsInvalid, ['E', '\\x7B2C']);

    // Volume
    add(ElementCategory.VolumePrefix, optionsDefault, ['VOL', 'VOL.', 'VOLUME']);

    // Release
    add(ElementCategory.ReleaseGroup, optionsDefault, ['Baha', 'THORA']);
    add(ElementCategory.ReleaseInformation, optionsDefault, [
      'BATCH',
      'COMPLETE',
      'PATCH',
      'REMUX'
    ]);
    add(ElementCategory.ReleaseInformation, optionsUnidentifiable, ['END', 'FINAL']);
    add(ElementCategory.ReleaseVersion, optionsDefault, [
      'V0',
      'V1',
      'V2',
      'V3',
      'V4',
      'V5',
      'V6',
      'V7',
      'V8',
      'V9'
    ]);

    // Other
    add(ElementCategory.Other, optionsDefault, [
      'REMASTER',
      'REMASTERED',
      'UNCENSORED',
      'UNCUT',
      'TS',
      'VFR',
      'WIDESCREEN',
      'WS'
    ]);

    // File extension
    add(ElementCategory.FileExtension, optionsDefault, [
      '3GP',
      'AVI',
      'DIVX',
      'FLV',
      'M2TS',
      'MKV',
      'MOV',
      'MP4',
      'MPG',
      'OGM',
      'RM',
      'RMVB',
      'TS',
      'WEBM',
      'WMV'
    ]);
    add(ElementCategory.FileExtension, optionsInvalid, [
      'AAC',
      'AIFF',
      'FLAC',
      'M4A',
      'MP3',
      'MKA',
      'OGG',
      'WAV',
      'WMA',
      '7Z',
      'RAR',
      'ZIP',
      'ASS',
      'SRT'
    ]);

    function add(category: ElementCategory, options: KeywordOptions, input: string[]) {
      const map = category === 'extension' ? extensions : keys;
      for (const key of input) {
        if (!map.has(key)) {
          map.set(key, { category, ...options });
        }
      }
    }
  }

  public static normalize(text: string) {
    return text.toLocaleUpperCase();
  }

  public static contains(category: ElementCategory, keyword: string) {
    const map = this.container(category);
    const value = map.get(keyword);
    return value && value.category === category;
  }

  public static container(category: ElementCategory) {
    return category === 'extension' ? this.extensions : this.keys;
  }

  public static find(keyword: string, category: ElementCategory) {
    const map = this.container(category);
    if (!map.has(keyword)) {
      return undefined;
    }
    const entry = map.get(keyword)!;
    if (category !== 'unknown' && entry.category !== category) {
      return undefined;
    }
    return entry;
  }

  public static peek(range: TextRange) {
    const search = range.toString();
    const result: ParsedResult = {};
    const predefined: TextRange[] = [];
    for (const { category, list } of this.peekEntries) {
      for (const key of list) {
        // StringComparison.CurrentCulture
        // https://learn.microsoft.com/ja-jp/dotnet/api/system.stringcomparison?view=net-7.0
        const foundIdx = search.indexOf(key);
        if (foundIdx === -1) continue;
        result[category] = key;
        predefined.push(range.fork(foundIdx + range.offset, key.length));
      }
    }
    return { result, predefined };
  }
}

export const Fansubs = new Set([
  '猎户发布组',
  '猎户手抄部',
  '北宇治字幕组',
  '北宇治Anarchism字幕组',
  '動漫花園',
  '拨雪寻春',
  'NC-Raws',
  '喵萌奶茶屋',
  'Lilith-Raws',
  '魔星字幕团',
  '桜都字幕组',
  '天月動漫&發佈組',
  '极影字幕社',
  'LoliHouse',
  '悠哈C9字幕社',
  '幻月字幕组',
  '天使动漫论坛',
  '动漫国字幕组',
  '幻樱字幕组',
  '爱恋字幕社',
  'DBD制作组',
  'c.c动漫',
  '萝莉社活动室',
  '千夏字幕组',
  'IET字幕組',
  '诸神kamigami字幕组',
  '霜庭云花Sub',
  'GMTeam',
  '风车字幕组',
  // '雪飄工作室(FLsnow)',
  'MCE汉化组',
  '丸子家族',
  '星空字幕组',
  '梦蓝字幕组',
  'LoveEcho!',
  'SweetSub',
  '枫叶字幕组',
  'Little Subbers!',
  '轻之国度',
  '云光字幕组',
  '豌豆字幕组',
  '驯兽师联盟',
  '中肯字幕組',
  'SW字幕组',
  '银色子弹字幕组',
  '风之圣殿',
  'YWCN字幕组',
  'KRL字幕组',
  '华盟字幕社',
  '波洛咖啡厅',
  '动音漫影',
  'VCB-Studio',
  'DHR動研字幕組',
  '80v08',
  '肥猫压制',
  'Little字幕组',
  'AI-Raws',
  '离谱Sub',
  '虹咲学园烤肉同好会',
  'ARIA吧汉化组',
  '柯南事务所',
  '百冬練習組',
  '冷番补完字幕组',
  '爱咕字幕组',
  '極彩字幕组',
  'AQUA工作室',
  '未央阁联盟',
  '届恋字幕组',
  '夜莺家族',
  'TD-RAWS',
  '夢幻戀櫻',
  'WBX-SUB',
  'Liella!の烧烤摊',
  'Amor字幕组',
  'MingYSub',
  '小白GM',
  'Sakura',
  'EMe',
  'Alchemist',
  '黑岩射手吧字幕组',
  'ANi',
  'MSB制作組'
]);
