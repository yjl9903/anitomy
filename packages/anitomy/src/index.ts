import type { AnitomyOptions, AnitomyResult, ParsedResult } from './types';

import { mergeResult } from './utils';
import { KeywordManager } from './keyword';
import { ElementCategory } from './element';
import { parse as doParse } from './parser';
import { tokenize as doTokenize } from './tokenizer';

export type * from './types';

/**
 * Parser instance with shared options and parsed result cache
 */
export class Parser {
  private readonly options: AnitomyOptions;

  private readonly cache = new Map<string, AnitomyResult | null>();

  public constructor(options: Partial<AnitomyOptions> = {}) {
    this.options = resolveOptions(options);
  }

  public parse(filename: string, force = false) {
    if (!force && this.cache.has(filename)) {
      const result = this.cache.get(filename);
      return result ? result : undefined;
    } else {
      const result = parse(filename, this.options);
      if (result) {
        this.cache.set(filename, result);
      } else {
        this.cache.set(filename, null);
      }
      return result;
    }
  }
}

export function parse(
  filename: string,
  _options: Partial<AnitomyOptions> = {}
): AnitomyResult | undefined {
  if (filename === '') return undefined;

  let result: ParsedResult = {};

  const options = resolveOptions(_options);

  result.filename = filename;
  if (options.parseFileExtension) {
    const ext = removeExtension(filename);
    if (ext) {
      result.filename = filename;
      result.extension = ext.extension;
    }
  }

  const tokenized = doTokenize(result.filename!, options);
  result = mergeResult(result, tokenized.result);
  if (!tokenized.ok) {
    return resolveResult(result);
  }

  const parsed = doParse(result, tokenized.tokens, options);
  result = parsed.result;

  return resolveResult(result);
}

export function resolveOptions(options: Partial<AnitomyOptions>): AnitomyOptions {
  return {
    delimiters: ' _.&+,|',
    parseEpisodeNumber: true,
    parseEpisodeTitle: true,
    parseFileExtension: true,
    parseReleaseGroup: true,
    ...options
  };
}

function resolveResult(result: ParsedResult): AnitomyResult {
  const resolved: AnitomyResult = {
    title: result['title'],
    type: result['type'],
    season: normalizeSeason(result['season']),
    year: normalizeNumber(result['year']),
    month: normalizeNumber(result['month']),
    language: result['language'],
    subtitles: result['subtitles'],
    source: result['source'],
    episode: {
      number: normalizeNumber(result['episode.number']),
      numberAlt: normalizeNumber(result['episode.numberAlt']),
      title: result['episode.title']
    },
    volume: {
      number: normalizeNumber(result['volume'])
    },
    video: {
      term: result['video.term'],
      resolution: result['video.resolution']
    },
    audio: {
      term: result['audio.term']
    },
    release: {
      version: normalizeNumber(result['release.version']),
      group: result['release.group']
    },
    file: {
      name: result['filename']!,
      extension: result['extension'],
      checksum: result['checksum']
    }
  };
  return resolved;

  function normalizeSeason(num: string | undefined) {
    if (num !== undefined && num !== null) {
      return /^\d+$/.test(num) ? String(+num) : num;
    } else {
      return undefined;
    }
  }

  function normalizeNumber(num: string | undefined) {
    try {
      if (num !== undefined && num !== null) {
        const n = parseFloat(num);
        return !Number.isNaN(n) ? n : undefined;
      }
    } catch {}
    return undefined;
  }
}

function removeExtension(filename: string) {
  const position = filename.lastIndexOf('.');
  if (position === -1) return undefined;

  const extension = filename.slice(position + 1);
  if (extension.length > 4 || !/^[a-zA-Z0-0]+$/.test(extension)) {
    return undefined;
  }

  if (!KeywordManager.contains(ElementCategory.FileExtension, extension)) {
    return undefined;
  }

  return {
    filename: filename.slice(0, position),
    extension
  };
}
