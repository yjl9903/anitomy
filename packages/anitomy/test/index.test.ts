import { describe, it, expect } from 'vitest';

import { parse } from '../src';

import { other, fails } from './testcase';

describe('Other filenames', () => {
  for (const filename of other) {
    it(`should parse ${filename}`, () => {
      const info = parse(filename);
      expect(info).toMatchSnapshot();
      expect(info?.title).toBeTruthy();
    });
  }

  for (const filename of fails) {
    it(`should parse not ${filename}`, () => {
      const info = parse(filename);
      expect(info).toMatchSnapshot();
      expect(info?.title).toBeFalsy();
    });
  }
});
