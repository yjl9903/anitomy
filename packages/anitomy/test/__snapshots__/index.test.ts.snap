// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Other filenames > should parse 1`] = `
{
  "audio": {
    "term": undefined,
  },
  "episode": {
    "number": 1080,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": undefined,
    "name": "[银色子弹字幕组][名侦探柯南][第1080集 对准灰原的相机][简繁日多语MKV][1080P]",
  },
  "language": undefined,
  "month": undefined,
  "release": {
    "group": "银色子弹字幕组",
    "version": undefined,
  },
  "season": undefined,
  "source": undefined,
  "subtitles": undefined,
  "title": "名侦探柯南",
  "type": undefined,
  "video": {
    "resolution": "1080P",
    "term": undefined,
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`Other filenames > should parse 2`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 2,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": "MP4",
    "name": "[Lilith-Raws] 熊熊勇闖異世界 PUNCH！ / Kuma Kuma Kuma Bear S02 - 02 [Baha][WEB-DL][1080p][AVC AAC][CHT][MP4]",
  },
  "language": "CHT",
  "month": undefined,
  "release": {
    "group": "Lilith-Raws",
    "version": undefined,
  },
  "season": "2",
  "source": "WEB-DL",
  "subtitles": undefined,
  "title": "熊熊勇闖異世界 PUNCH！ / Kuma Kuma Kuma Bear S02",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "AVC",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`Other filenames > should parse 3`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 13,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": "MP4",
    "name": "[ANi] 機動戰士鋼彈 水星的魔女 Season 2 - 13 [1080P][Baha][WEB-DL][AAC AVC][CHT][MP4]",
  },
  "language": "CHT",
  "month": undefined,
  "release": {
    "group": "ANi",
    "version": undefined,
  },
  "season": "2",
  "source": "WEB-DL",
  "subtitles": undefined,
  "title": "機動戰士鋼彈 水星的魔女",
  "type": undefined,
  "video": {
    "resolution": "1080P",
    "term": "AVC",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`Other filenames > should parse 4`] = `
{
  "audio": {
    "term": undefined,
  },
  "episode": {
    "number": 1,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": "MP4",
    "name": "[爱恋字幕社][4月新番][魔法使的新娘2][Mahoutsukai no Yome S2][01][1080P][MP4][GB][简中]",
  },
  "language": "简中",
  "month": 4,
  "release": {
    "group": "爱恋字幕社",
    "version": undefined,
  },
  "season": "2",
  "source": undefined,
  "subtitles": "GB",
  "title": "魔法使的新娘2",
  "type": undefined,
  "video": {
    "resolution": "1080P",
    "term": undefined,
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`Other filenames > should parse 5`] = `
{
  "audio": {
    "term": undefined,
  },
  "episode": {
    "number": 1,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": undefined,
    "name": "[桜都字幕组] 总之就是非常可爱 S2 / Tonikaku Kawaii S2 [01][1080p][简体内嵌]",
  },
  "language": undefined,
  "month": undefined,
  "release": {
    "group": "桜都字幕组",
    "version": undefined,
  },
  "season": "2",
  "source": undefined,
  "subtitles": undefined,
  "title": "总之就是非常可爱 S2 / Tonikaku Kawaii S2",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": undefined,
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`Other filenames > should parse 6`] = `
{
  "audio": {
    "term": undefined,
  },
  "episode": {
    "number": 1,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": undefined,
    "name": "【喵萌奶茶屋】★04月新番★[百合是我的工作！/我的百合乃工作是也！/私の百合はお仕事です！/Watashi no Yuri wa Oshigoto desu!][01][720p][繁日雙語][招募翻譯校對]",
  },
  "language": undefined,
  "month": 4,
  "release": {
    "group": "喵萌奶茶屋",
    "version": undefined,
  },
  "season": undefined,
  "source": undefined,
  "subtitles": undefined,
  "title": "百合是我的工作！/我的百合乃工作是也！/私の百合はお仕事です！/Watashi no Yuri wa Oshigoto desu!",
  "type": undefined,
  "video": {
    "resolution": "720p",
    "term": undefined,
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`Other filenames > should parse 7`] = `
{
  "audio": {
    "term": undefined,
  },
  "episode": {
    "number": 13,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": "MP4",
    "name": "【幻樱字幕组】【7月新番】【异世界舅舅 Isekai Ojisan】【13_END】【BIG5_MP4】【1920X1080】",
  },
  "language": undefined,
  "month": 7,
  "release": {
    "group": "幻樱字幕组",
    "version": undefined,
  },
  "season": undefined,
  "source": undefined,
  "subtitles": "BIG5",
  "title": "异世界舅舅 Isekai Ojisan",
  "type": undefined,
  "video": {
    "resolution": "1920X1080",
    "term": undefined,
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`Other filenames > should parse 8`] = `
{
  "audio": {
    "term": undefined,
  },
  "episode": {
    "number": 13,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": "MP4",
    "name": "[爱恋字幕社&漫猫字幕社][4月新番][和山田进行LV999 的恋爱][Yamada-kun to Lv999 no Koi wo Suru][13Fin][1080p][MP4][简中]",
  },
  "language": "简中",
  "month": 4,
  "release": {
    "group": "漫猫字幕社",
    "version": undefined,
  },
  "season": undefined,
  "source": undefined,
  "subtitles": undefined,
  "title": "和山田进行LV999 的恋爱",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": undefined,
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`Other filenames > should parse 9`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 1,
    "numberAlt": 13,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": "MP4",
    "name": "[爱恋字幕社&漫猫字幕社] 和山田进行LV999 的恋爱/Yamada-kun to Lv999 no Koi wo Suru (01-13Fin WEBRIP 720p AVC AAC MP4 2023年4月 简中)",
  },
  "language": "简中",
  "month": undefined,
  "release": {
    "group": "漫猫字幕社",
    "version": undefined,
  },
  "season": undefined,
  "source": "WEBRIP",
  "subtitles": undefined,
  "title": "和山田进行LV999 的恋爱/Yamada-kun to Lv999 no Koi wo Suru",
  "type": undefined,
  "video": {
    "resolution": "720p",
    "term": "AVC",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`Other filenames > should parse 10`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": undefined,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": "08FF8DBF",
    "extension": "mp4",
    "name": "[DMG&RoxyLib] 無職転生 [BDRip][1080P][CHS][MP4]/[DMG&RoxyLib] 無職転生 番外編「エリスのゴブリン討伐」[BDRip][AVC_AAC][1080P][CHS](08FF8DBF).mp4",
  },
  "language": "CHS",
  "month": undefined,
  "release": {
    "group": undefined,
    "version": undefined,
  },
  "season": undefined,
  "source": "BDRip",
  "subtitles": undefined,
  "title": "無職転生",
  "type": "番外編",
  "video": {
    "resolution": "1080P",
    "term": "AVC",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`Other filenames > should parse 11`] = `
{
  "audio": {
    "term": undefined,
  },
  "episode": {
    "number": 1,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": "MP4",
    "name": "【极影字幕社】★7月新番 白圣女与黑牧师 Shiro Seijo to Kuro Bokushi 第01v2话 GB 1080P MP4（字幕社招人内详）",
  },
  "language": undefined,
  "month": 7,
  "release": {
    "group": "极影字幕社",
    "version": 2,
  },
  "season": undefined,
  "source": undefined,
  "subtitles": "GB",
  "title": "白圣女与黑牧师 Shiro Seijo to Kuro Bokushi",
  "type": undefined,
  "video": {
    "resolution": "1080P",
    "term": undefined,
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`Other filenames > should parse 12`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": undefined,
    "numberAlt": undefined,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": "MP4",
    "name": "[ANi] 無職轉生～到了異世界就拿出真本事 第二季 - 特別篇 [1080P][Baha][WEB-DL][AAC AVC][CHT][V2][MP4]",
  },
  "language": "CHT",
  "month": undefined,
  "release": {
    "group": "ANi",
    "version": 2,
  },
  "season": "2",
  "source": "WEB-DL",
  "subtitles": undefined,
  "title": "無職轉生～到了異世界就拿出真本事 第二季 - 特別篇",
  "type": "特別篇",
  "video": {
    "resolution": "1080P",
    "term": "AVC",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`Other filenames > should parse 13`] = `
{
  "audio": {
    "term": "AAC",
  },
  "episode": {
    "number": 1,
    "numberAlt": 12,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": undefined,
    "name": "[LoliHouse] 满怀美梦的少年是现实主义者 / Yumemiru Danshi wa Genjitsushugisha [01-12合集][WebRip 1080p HEVC-10bit AAC][简繁内封字幕][Fin]",
  },
  "language": undefined,
  "month": undefined,
  "release": {
    "group": "LoliHouse",
    "version": undefined,
  },
  "season": undefined,
  "source": "WebRip",
  "subtitles": undefined,
  "title": "满怀美梦的少年是现实主义者 / Yumemiru Danshi wa Genjitsushugisha",
  "type": undefined,
  "video": {
    "resolution": "1080p",
    "term": "HEVC-10bit",
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;

exports[`Other filenames > should parse 14`] = `
{
  "audio": {
    "term": undefined,
  },
  "episode": {
    "number": 1,
    "numberAlt": 12,
    "title": undefined,
  },
  "file": {
    "checksum": undefined,
    "extension": "MP4",
    "name": "[轻之国度字幕组][女友成双/女朋友 and 女朋友][第01-12话][720P][MP4]",
  },
  "language": undefined,
  "month": undefined,
  "release": {
    "group": "轻之国度字幕组",
    "version": undefined,
  },
  "season": undefined,
  "source": undefined,
  "subtitles": undefined,
  "title": "女友成双/女朋友 and 女朋友",
  "type": undefined,
  "video": {
    "resolution": "720P",
    "term": undefined,
  },
  "volume": {
    "number": undefined,
  },
  "year": undefined,
}
`;
